http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2MultiplexedChannelPool.java

```diff
@@ -46,151 +58,333 @@
+                // If we don't have a multiplexed channel, the parent channel hasn't been fully initialized. Close it now.
+                closeAndReleaseParent(parentChannel);
+            }
+        } catch (Exception e) {
+            log.error(() -> "Failed to handle GOAWAY frame on channel " + parentChannel, e);
+        }
     }
 
     @Override
     public void close() {
-        try {
-            setClosedFlag().await();
-            for (MultiplexedChannelRecord c : connections) {
-                Future<Channel> f = c.getConnectionFuture();
-                f.await();
-                if (f.isSuccess()) {
-                    connectionPool.release(f.getNow()).await();
+        if (closed.compareAndSet(false, true)) {
+            Future<?> closeCompleteFuture = doClose();
+
+            try {
+                if (!closeCompleteFuture.await(10, TimeUnit.SECONDS)) {
+                    throw new RuntimeException("Event loop didn't close after 10 seconds.");
                 }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            }
+
+            Throwable exception = closeCompleteFuture.cause();
+            if (exception != null) {
+                throw new RuntimeException("Failed to close channel pool.", exception);
             }
-            connectionPool.close();
-        } catch (InterruptedException ie) {
-            Thread.currentThread().interrupt();
-            throw new RuntimeException(ie);
         }
     }
 
-    private Promise<Void> setClosedFlag() {
-        Promise<Void> closedFuture = eventLoop.newPromise();
-        doInEventLoop(eventLoop, () -> {
-            closed = true;
-            closedFuture.setSuccess(null);
+    private Future<?> doClose() {
+        EventLoop closeEventLoop = eventLoopGroup.next();
+        Promise<?> closeFinishedPromise = closeEventLoop.newPromise();
+
+        doInEventLoop(closeEventLoop, () -> {
+            Promise<Void> releaseAllChannelsPromise = closeEventLoop.newPromise();
+            PromiseCombiner promiseCombiner = new PromiseCombiner(closeEventLoop);
+
+            // Create a copy of the connections to remove while we close them, in case closing updates the original list.
+            List<MultiplexedChannelRecord> channelsToRemove = new ArrayList<>(connections);
+            for (MultiplexedChannelRecord channel : channelsToRemove) {
+                promiseCombiner.add(closeAndReleaseParent(channel.getConnection()));
+            }
+            promiseCombiner.finish(releaseAllChannelsPromise);
+
+            releaseAllChannelsPromise.addListener(f -> {
+                connectionPool.close();
+                closeFinishedPromise.setSuccess(null);
+            });
         });
-        return closedFuture;
+
+        return closeFinishedPromise;
+    }
+
+    private final class ReleaseOnExceptionHandler extends ChannelDuplexHandler {
```

Comment:
We can use `@Sharable`.

Since handler accesses private methods, I'd prefer to keep it internal. Communication between the two would be weird, otherwise.